### DISCUSSION TITLE
Migration from Xtext to Langium

### DISCUSSION THREAD

**Discussion by ballcoach12** [eclipse-langium/langium#212] [Q&A] [answered]
Hi Guys,

My team of developers are long-time (5+ years) users of the Xtext/Xtend/Xcore ecosystem. We are currently using it to:

- Create independent domain models
- Create grammars, lexers, parsers, scoping, validation, and code generation assets that are based on the domain models
- Create Java back ends with a common messaging system that supports the handling of diagram-centric messages for CRUD operations related to visual design tools. The back end translates the messages into the AST that is linked to the domain model and is validated based on domain model rules.
- Create web-based front ends that communicate with the back end using our diagram-centric messages; we also occasionally use LSP for text-based editors that are either standalone or work simultaneously with the graphical editor.

We would like to consider Langium as a possible Xtext replacement at some point, and we have some questions.

Based on what I've seen of Langium, your goal is to remove the complexity of things like EMF/JVM/Eclipse and provide a simple solution that supports VS Code extensions for editors along with the ability to generate code for a language server. Is this a fair characterization of the project goals?

Will there be any support for syntactically defining independent domain models that can be plugged into grammars? (it appears that Langium only supports the generation of types from the grammar at the present time).

We are accustomed to DI using Guice, having come from the Xtext/Java world. Is there a DI mechanism that is supported in Langium? Is there even a need for DI with Langium?

Can you elaborate on how scoping and serialization work in Langium? I am accustomed to writing ScopeProviders for Xtext. 

What type of generator is/will be supported? We have a strong need to generate code and to M2M from our domain models so that we can represent the model in various target formats.

Thanks in advance,

Greg Lusk

**Comment by msujew** [comment]
Hi @ballcoach12, thank you for your interest in Langium at this early stage :)

> Is this a fair characterization of the project goals?

Yes. Our main goal is to provide feature parity to Xtext in regards to their language server feature, without all of the overhead that Xtext brings.

> Will there be any support for syntactically defining independent domain models that can be plugged into grammars?

Yes, that is planned to be supported in the near future. Discussion on that and implementation proposals are happening in #91.

> Is there a DI mechanism that is supported in Langium?

Yes, we have implemented our small custom DI framework in [this file](https://github.com/langium/langium/blob/main/packages/langium/src/dependency-injection.ts). You can see how one creates their [own injection module here](https://github.com/langium/langium/blob/main/packages/langium/src/default-module.ts) and an example how it it is [used here](https://github.com/langium/langium/blob/dc89de7a271af14366c47ceaada052efd67cc3bc/packages/langium/src/lsp/reference-finder.ts#L20-L27). We will soon provide a documentation/tutorials page which will contain more info on the DI mechanism.

> I am accustomed to writing ScopeProviders for Xtext.

Well, then you're in luck. We didn't really see a lot of reason to change how scoping works in Langium compared to Xtext. The main difference is how the default scoping provider is implemented: The default in Xtext provides a static access to all elements everywhere using fully qualified names, while we try to keep it more simple. Instead, we opted for a simple local scope based scoping as the default that is more commonly used in programming languages. However, nothing stops anyone from implementing something similar to Xtext, which we actually did for our [`DomainModel` example](https://github.com/langium/langium/blob/main/examples/domainmodel/src/language-server/domain-model-scope.ts).

> What type of generator is/will be supported?

We try to align with the conventions of the JS ecosystem, therefore our recommended solution is to build a CLI for generation. Our own [`langium-cli`](https://www.npmjs.com/package/langium-cli) is built that way. We plan to provide utilities for writing those, in particular help for watch modes, file-access, cli-output, etc. You could hook a generator into our `DocumentBuilder`, but that might lead to performance issues down the line: Compared to a JVM language we are limited to using node.js' event loop on a single thread.

**Reply by ballcoach12** [reply]
@msujew, thank you for the quick reply. This is very exciting news! We would be interested in providing input to guide features and even contributing code if there is a need. 

We are looking forward to seeing how this project progresses!

Greg

**Reply by trusktr** [reply]
> without all of the overhead that Xtext brings.

Can you describe what are the overheads? (I've never used Xtext or made a language before, and have been having fun with Langium so far! Want to know when to use one vs the other)

I can already see that working with a TypeScript/Node.js setup is a lot easier, and attractive for web developers (lots of web developers). That's a big win!

**Reply by msujew** [reply]
Xtext is based on the [Eclipse Modelling Framework](https://www.eclipse.org/modeling/emf/) (EMF). It's a fairly heavyweight approach to model driven development and Xtext is tightly integrated into the framework. It comes with a lot of utility for Xtext, since it already proposes an approach to designing meta models that Xtext simply reuses. However, with EMF also comes a steep learning curve that makes Xtext also difficult to approach as a newcomer.

With Langium, we decided that we don't need some underlying modelling framework. We just roll with some generated interfaces and a very simplistic reflection object that allows us to inspect these interfaces during runtime. That makes Langium way more approachable, which was one of our initial goal when we started development on it.

**Comment by dhuebner** [comment, marked as answer]
@ballcoach12
Maybe this MWE2 fragment could be interesting for you:
https://github.com/TypeFox/xtext2langium#readme

**Reply by ballcoach12** [reply]
Oh wow! That's huge! Thanks for sharing. This could open up a lot of possibilities for my team!

**Reply by steve-hickman-epistimis** [reply]
I was excited to use this as well - but I ran into my own ignorance. See [this issue](https://github.com/TypeFox/xtext2langium/issues/1).  Could someone educate me by telling me what I missed?

**Reply by spoenemann** [reply]
Yes classpath issues can be painful in Java. The solution highly depends on your build setup, so it's hard to give general hints about that.
